jupyter/d2l-java/chapter_multilayer-perceptrons/mlp-djl.ipynb

```diff
@@ -0,0 +1,311 @@
+    "net.add(Activation::relu);\n",
+    "net.add(Linear.builder().setOutChannels(10).build());\n",
+    "net.setInitializer(new NormalInitializer());"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that DJL, as usual, automatically\n",
+    "infers the missing input dimensions to each layer.\n",
+    "\n",
+    "The training loop is *exactly* the same\n",
+    "as when we implemented softmax regression.\n",
+    "This modularity enables us to separate \n",
+    "matters concerning the model architecture\n",
+    "from orthogonal considerations."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int batchSize = 256;\n",
+    "int numEpochs = 10;\n",
+    "float[] trainLoss;\n",
+    "float[] testAccuracy;\n",
+    "float[] epochCount;\n",
+    "float[] trainAccuracy;\n",
+    "\n",
+    "FashionMnist trainIter = FashionMnist.builder()\n",
+    "                            .optUsage(Dataset.Usage.TRAIN)\n",
+    "                            .setSampling(batchSize, true)\n",
+    "                            .build();\n",
+    "\n",
+    "\n",
+    "FashionMnist testIter = FashionMnist.builder()\n",
+    "                            .optUsage(Dataset.Usage.TEST)\n",
+    "                            .setSampling(batchSize, true)\n",
+    "                            .build();\n",
+    "\n",
+    "trainIter.prepare();\n",
+    "testIter.prepare();\n",
+    "\n",
+    "for(int i = 0; i < epochCount.length; i++) {\n",
+    "    epochCount[i] = (i + 1);\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "6"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "LearningRateTracker lrt = LearningRateTracker.fixedLearningRate(0.5f);\n",
+    "Optimizer sgd = Optimizer.sgd().setLearningRateTracker(lrt).build();\n",
+    "\n",
+    "Loss loss = Loss.softmaxCrossEntropyLoss();\n",
+    "\n",
+    "DefaultTrainingConfig config = new DefaultTrainingConfig(loss)\n",
+    "                .optOptimizer(sgd) // Optimizer (loss function)\n",
+    "                .addEvaluator(new Accuracy()) // Model Accuracy\n",
+    "                .addTrainingListeners(TrainingListener.Defaults.basic()); // Logging\n",
+    "\n",
+    "    try (Model model = Model.newInstance(\"mlp\")) {\n",
+    "        model.setBlock(net);\n",
+    "\n",
+    "        try (Trainer trainer = model.newTrainer(config)) {\n",
+    "\n",
+    "            trainer.initialize(new Shape(1, 784));\n",
+    "            trainer.setMetrics(new Metrics());\n",
+    "\n",
+    "            EasyTrain.fit(trainer, numEpochs, trainIter, testIter);\n",
+    "\n",
+    "            Metrics metrics = trainer.getMetrics();\n",
+    "\n",
+    "            for (Evaluator evaluator : trainer.getEvaluators()) {\n",
```

Comment:
This code for using the metrics is quite large and might distract from the ideas you are really trying to get across in the section. See if you can simplify it. The Java 8 streaming API might help. If you feel like the DJL API is lacking, go ahead and raise a PR to improve it

Reply:
Would you want to reduce the Evaluator loop itself or the loops inside it ? 
Also, it might have a problem but streaming API would be able to turn the testAccuracy, trainAccuracy to Float[] and not float[] which would cause problems with the table saw api.

Reply:
I have made a few changes to this piece of code to make it look cleaner. I have made changes to only mlp-djl.ipynb file specifically for now. Let me know if I can further improve it. I will extend these changes to other files if it seems alright.